### PR TITLE
Minor refactor of fully connected.

### PR DESCRIPTION
**Pull request by larroy**
## Description ##

This factors out code in a separate PR as requested here: https://github.com/apache/incubator-mxnet/pull/14779

The refactor extracts logic to get the product shapes for the two fully
connected branches, flattened and unflattened inputs.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by larroy** [comment]
@apeforest

**Comment by abhinavs95** [comment]
@mxnet-label-bot add [pr-awaiting-review]

File: src/operator/nn/fully_connected-inl.h [hunk 1]
```diff
@@ -75,6 +75,38 @@ struct FullyConnectedParam : public dmlc::Parameter<FullyConnectedParam> {
...
+/**
+ * Flatten dimensions except last
+ * @tparam xpu
+ * @tparam DType
+ * @param tblob
+ * @param ctx
+ * @return 2 Dimensional tensor with front shapes collapsed
+ */
+template<typename xpu, typename DType>
+Tensor<xpu, 2, DType> FlattenAs2DHead(const TBlob& tblob, const OpContext& ctx) {
```
Comment (apeforest) [inline src/operator/nn/fully_connected-inl.h]: These two methods seem to be generic for all operators. Can we add them to a common place?

File: src/operator/nn/fully_connected-inl.h [hunk 2]
```diff
@@ -134,59 +159,52 @@ void FCBackward(const OpContext &ctx, const FullyConnectedParam &param,
...
-  const mxnet::TShape& ishape = in_data[fullc::kData].shape_;
-  const mxnet::TShape& oshape = out_grad[fullc::kOut].shape_;
-
-  Tensor<xpu, 2, DType> wmat = in_data[fullc::kWeight].get<xpu, 2, DType>(s);
-  Tensor<xpu, 2, DType> data, grad, gdata;
+  Stream<xpu> *stream = ctx.get_stream<xpu>();
+  Tensor<xpu, 2, DType> wmat = in_data[fullc::kWeight].get<xpu, 2, DType>(stream);
+  Tensor<xpu, 2, DType> x, y_grad, x_grad;
   if (!param.flatten) {
-    data = in_data[fullc::kData].get_with_shape<xpu, 2, DType>(
```
Comment (apeforest) [inline src/operator/nn/fully_connected-inl.h]: These variable names are used in many operators. Why refactor them?

**Comment by larroy** [comment]
You requested this refactor as a separate PR on FC 2nd order gradient PR. I'm a bit confused by your review feedback. I'm not sure what you mean or what changes are you suggesting.

**Comment by larroy** [comment]
https://github.com/apache/incubator-mxnet/pull/14779

File: src/operator/nn/fully_connected-inl.h (hunk 1 above)
Comment (larroy) [reply src/operator/nn/fully_connected-inl.h]: Please suggest a better name.

File: src/operator/nn/fully_connected-inl.h (hunk 2 above)
Comment (larroy) [reply src/operator/nn/fully_connected-inl.h]: I don't think data is a good variable name. Are you against this rename? Could you be specific?

**Comment by larroy** [comment]
The repeated logic is factored out for readability, does every refactor have to be justified with performance impact? Maintainability and code clarity are also important. If you look at the second order gradient code, this functionality is repeated multiple times, so it makes sense to factor out in a separate function. Are you against this principle or what is the cause for the negative unspecific feedback on this PR? Are you suggesting the PR should be closed? You asked specifically to make a separate PR for this change.

File: src/operator/nn/fully_connected-inl.h (hunk 2 above)
Comment (larroy) [reply src/operator/nn/fully_connected-inl.h]: I can rename back to data if that's the blocker.

**Pull request closed without merging**